connectivity/mqtt/duplex_utils.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "duplex_utils.h"
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+status_t duplex_config_init(struct mosquitto **config_mosq, mosq_config_t *config_cfg) {
+  status_t ret = SC_OK;
+  if (config_mosq == NULL || config_cfg == NULL) {
+    return SC_MQTT_NULL;
+  }
+
+  init_mosq_config(config_cfg, client_duplex);
+  mosquitto_lib_init();
+
+  ret = generate_client_id(config_cfg);
+  if (ret) {
+    return ret;
+  }
+
+  init_check_error(config_cfg, client_pub);
+
+  *config_mosq = mosquitto_new(config_cfg->general_config->id, true, NULL);
+  if (!config_mosq) {
+    switch (errno) {
+      case ENOMEM:
+        fprintf(stderr, "Error: Out of memory.\n");
+        break;
+      case EINVAL:
+        fprintf(stderr, "Error: Invalid id.\n");
+        break;
+    }
+    return SC_MOSQ_OBJ_INIT_ERROR;
+  }
+
+  ret = mosq_opts_set(*config_mosq, config_cfg);
+
+  return ret;
+}
+
+status_t gossip_channel_set(mosq_config_t *channel_cfg, char *host, char *sub_topic, char *pub_topic) {
+  status_t ret = SC_OK;
+  if (channel_cfg == NULL || (host == NULL && sub_topic == NULL && pub_topic == NULL)) {
+    return SC_MQTT_NULL;
+  }
+
+  if (host) {
+    channel_cfg->general_config->host = strdup(host);
+    channel_cfg->general_config->client_type = client_pub;
+  }
+
+  if (sub_topic) {
+    ret = cfg_add_topic(channel_cfg, client_sub, sub_topic);
+    if (ret) {
+      goto done;
+    }
+  }
+
+  if (pub_topic) {
+    ret = cfg_add_topic(channel_cfg, client_pub, pub_topic);
+  }
+
+done:
+  return ret;
+}
+
+status_t gossip_api_channels_set(mosq_config_t *channel_cfg, char *host, char *root_path) {
+  status_t ret = SC_OK;
+
+  if (channel_cfg == NULL || host == NULL || root_path == NULL) {
+    return SC_MQTT_NULL;
+  }
+
+  channel_cfg->general_config->host = strdup(host);
+  channel_cfg->general_config->client_type = client_pub;
+
+  char *sub_topic = NULL;
+  int sub_topic_len, api_name_len;
+  int root_path_len = strlen(root_path);
+  char *api_names[API_NUM] = {"generate_address", "get_tips_pair", "get_tips",          "receive_mam_message",
+                              "mam_send_message", "send_transfer", "find_transactions", "find_transaction_objects",
+                              "send_trytes"};
+
+  for (int i = 0; i < API_NUM; i++) {
+    api_name_len = strlen(api_names[i]);
+    sub_topic_len = root_path_len + 1 + api_name_len;
+    sub_topic = (char *)malloc(sub_topic_len + 1);
```

Comment:
`sub_topic_len` has already added 1 in line 93, why do you need to add again in line 94?

Reply:
The added 1 in line 93 is for the slash between `root_path_len` and `api_name_len`, since the topic would look like `<root_path_len>/<api_name_len>`. And the added 1 in the line 94 is for `\0`